### ISSUE TITLE
Arena Testing

### ISSUE THREAD

**Issue by krakjoe** [krakjoe/parallel#82]
Hi all,

I'm preparing some major internal changes to parallel. To start off with, I've swapped the system allocator for a shared mapped memory based allocator.

I'm interested to know if anyone can find any performance differences when running your test code against the "arena" branch ?

If anyone could do numbers, that'd be good ...

**Comment by krakjoe** [comment]
Oh note, you can now limit the memory available to parallel (and must set a maximum) with `parallel.memory` ini option ... I hate INI too, but necessary evil ... 128mb default, suffixes supported.

**Comment by WyriHaximus** [comment]
> Oh note, you can now limit the memory available to parallel (and must set a maximum) with `parallel.memory` ini option ... I hate INI too, but necessary evil ... 128mb default, suffixes supported.

Is this per thread or in total? Does it behave like the core memory limit setting (and can it be set to unlimited)?

**Comment by krakjoe** [comment]
It can't be unlimited, and is global

Cheers

**Comment by dktapps** [comment]
To be clear, that would be 128MB maximum for all threads combined, right?

**Comment by krakjoe** [comment]
Right

**Comment by krakjoe** [comment]
I think maybe I didn't give you enough of a reason to test this ... 

The most substantial barrier to parallel becoming a core part of PHP, which I very much want, is the ZTS requirement.

The environment that ZTS provides (share nothing) can also be provided by forking, and this we can support in NTS mode ...

It requires that we use shared memory, which is why we started with a shared arena ...

Now, onto the reason for feedback ... obviously, I can write benchmarks and measure, however ... The allocator at the moment is very very simple, many optimizations are possible. APCu also has a similar kind of allocator (so does opcache), and it uses just about every optimization that could be thought of (opcache uses almost none), and it sometimes behaves strangely and the likes of Nikita are not able to figure out why. So I don't want to over-optimize the thing if it doesn't seem necessary for real world code ... Of course if it totally ruins real code then I'll work on it ...

This is no pipe dream, I have a copy working locally with failures (that I understand) ... the only questions left now are how long to spend on optimizing the allocator, and how much of the API will need to change in v2 to support NTS and ZTS seamlessly (for example, bootstrapping may need to be changed, and I don't want to store a list of global channels for Channel::open anymore) ...

**Comment by dktapps** [comment]
why exactly is the zts requirements a barrier to becoming a core extension?

personally I've been using zts for so long now that I can't envision having a problem with sticking to it. Besides I can't imagine that forking will be better in terms of efficiency than using pthread.

**Comment by krakjoe** [comment]
> why exactly is the zts requirements a barrier to becoming a core extension?

ZTS is not enabled by default because of performance degradation that is unavoidable. It's better in newer versions, but unavoidable whatever.

To support only ZTS mode makes the extension too specialised to be a candidate for inclusion in core.

> Besides I can't imagine that forking will be better in terms of efficiency than using pthread.

The normal reason you create a thread over a fork is that threads share memory by default, they don't have to create a copy of the address space. In PHP, that doesn't really matter, whether you have a thread or a process, it must be initialized in the same way. The only place you might find a significant difference in terms of improvement or degradation is in context switching, which can sometimes be measured to be cheaper for threads than processes.

Note that we still have to support ZTS mode, because that's all windows has, so it's not going anywhere ...

Supporting NTS just gives parallel more reach, and makes it a more suitable candidate for inclusion in core ...

I'll go further, the work going on, almost certainly has the potential to improve performance in ZTS (because we are not calling system allocator). NTS has the potential to be even faster than ZTS because once forked (which is not really different to creating a thread because of COW optimizations at the system level), the environment is the same, but during runtime, the thread is executing the efficient version of instructions to access module globals rather than the slow version required by ZTS that causes the ZTS perf degradation.

The potential for ZTS improvements may not be realised (or may already be in effect), I haven't measured ... I'm sorta waiting on y'all ...

TL;DR this could wind up being good for everyone ...

**Comment by dktapps** [comment]
I'm not sold on ditching global channels for ::open(), because I think those have some uses for static lazy initialization (I'd actually like to see the same thing for Sync). It would be nice if that can be preserved.

I don't have any real world code that uses parallel to test with right now (PM is going to be a big job to migrate).

**Comment by krakjoe** [comment]
@dktapps noted ... I'll try to keep it and think about named syncs at some time in the future ...

Sorry, I haven't had more time to work on this, maybe next weekend ... maybe some time this week ...

**Comment by dktapps** [comment]
take all the time you need, no rush :+1:

**Comment by Tuckbros** [comment]
Ok, I read to fast. It is preparatory work in order to be able get rid of the ZTS requirement. It is not yet done. But it will definitely help everybody using the php coming with there linux distribution if they don't have to build php to use parallel.

**Comment by danielspk** [comment]
Support for NTS versions would be excellent for popularizing the project. :+1:

**Comment by krakjoe** [comment]
This has been put on hold ... it's a hell of a lot of work to bring these ideas to completion ...

**Comment by krakjoe** [comment]
Realistically, this isn't going to happen.

If I don't reduce the maintenance burden, I can't continue to maintain this. Committing myself to things like this one while trying to reduce maintenance burden doesn't make sense.
